### PR TITLE
Implement the segment upload protocol

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
Added custom map in segment ZK metadata
Added query parameter for parallel push protection
Modified BaseSegmentConversionExecutor to enable push retry, set custom map properly, and enable parallel push protection for minion.

For refresh case, upload segment as following steps:
1. Check IF-MATCH CRC if existing
2. Lock the segment if parallel push protection enabled
3. Update the custom map in segment ZK metadata
4. Update the segment ZK metadata (also unlock the segment)

### CODE REVIEW

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/backfill/BackfillSegmentUtils.java [hunk 1]
```diff
@@ -168,17 +163,16 @@ public boolean uploadSegment(String segmentName, File segmentDir, File outputDir
     try {
       TarGzCompressionUtils.createTarGzOfDirectory(segmentDir.getAbsolutePath(), segmentTar.getAbsolutePath());
       LOGGER.info("Created tar of {} at {}", segmentDir.getAbsolutePath(), segmentTar.getAbsolutePath());
-      InputStream inputStream = new FileInputStream(segmentTar);
-      int status = FileUploadUtils.sendSegmentFile(_controllerHost, _controllerPort, segmentName, inputStream, segmentTar.length());
-      if (status != 200) {
+      int statusCode =
+          FileUploadUtils.uploadSegment(_controllerHost, Integer.parseInt(_controllerPort), segmentName, segmentTar);
+      if (statusCode != HttpStatus.SC_OK) {
```
Comment (jenniferdai) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/backfill/BackfillSegmentUtils.java]: how do we distinguish between transient and non-transient exceptions?

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/CommonConstants.java [hunk 2]
```diff
@@ -312,19 +311,16 @@ public ServerType getServerType() {
     public static final String CREATION_TIME = "segment.creation.time";
     public static final String FLUSH_THRESHOLD_SIZE = "segment.flush.threshold.size";
     public static final String PARTITION_METADATA = "segment.partition.metadata";
-    public static final String OPTIMIZATIONS = "segment.optimizations";
     /**
      * This field will be populated only for use cases that are refreshing segments and
      * have minion configured. It will contain COMMITTING + human readable timestamp in UTC, time right
      * before any permanent changes happen to the segment, aka, right before
      * the segment is moved to its permanent directory. For example. it will say something like
      * COMMITTING-20171210T09:31:55.765
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/com/linkedin/pinot/common/utils/CommonConstants.java]: Good catch

File: pinot-common/src/main/java/com/linkedin/pinot/serde/SerDe.java [hunk 3]
```diff
@@ -15,7 +15,7 @@
  */
 package com.linkedin.pinot.serde;
 
-import org.apache.http.annotation.NotThreadSafe;
+import javax.annotation.concurrent.NotThreadSafe;
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/com/linkedin/pinot/serde/SerDe.java]: The old one is incorrect (and with new library, it does not exist)

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/backfill/BackfillSegmentUtils.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-tools/src/main/java/com/linkedin/pinot/tools/backfill/BackfillSegmentUtils.java]: Client decides that based on the status code, and do retry (you can use the RetryPolicy)

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/CommonConstants.java [hunk 4]
```diff
@@ -312,19 +311,16 @@ public ServerType getServerType() {
...
      */
-    public static final String SEGMENT_PUSH_STATUS = "segment.push.status";
-    /**
-     * This field will be a map of minion task type : time task ran in millis.
-     */
-    public static final String TASK_SEGMENT_MODIFICATION_TIME = "task.segment.modification.time";
+    public static final String SEGMENT_UPLOAD_START_TIME = "segment.upload.start.time";
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/com/linkedin/pinot/common/utils/CommonConstants.java]: Discussed offline, and decide to use segment.upload.start.time to differentiate this with LLC commit

**Pull request merged**